### DISCUSSION TITLE
dedicated layers (fixed Stat + Geom)

### DISCUSSION THREAD

**Discussion by EvaMaeRey** [ggplot2-extenders/ggplot-extension-club#42] [General]
I think the aliasing (stat_() -> geom_()) move is probably better than what's self-quoted below for first intro to creating layers w/ geom_* prefix.  Below is current stat of cookbook which also has a geom_coordinates function, briefly 'we hardcode both GeomText and StatCoordinates because that makes sense to me':

![image](https://github.com/user-attachments/assets/ec888c29-20d7-46e3-84cc-e80ce860496e)

But, I'm feeling okay about writing locked in (Stat and Geom) and  user-facing functions.

`geom_new <- function(...){layer(geom = "point", stat = StatNew, .etc.)}`  or 
`geom_new <- function(...){ geom_point(stat = StatNew, ...)}`  

I think this kind of runs counter to a hadley quote that having stat_ and geom_ functions was the wrong way to go, as it's less free and rather than more free and more in line with bit by bit composition.  But maybe I don't totally understand his idea or the context. And also I don't think I could track down that admission and any further discussion.  I think most recently heard from @yjunechoe in ggtrace talk...

One downside is there's a chance that someone erroneously thinks there's a GeomNew to work with. It's actually happened to me with ggbump::geom_sigmoid.  

Dedicated layers, vs. just creating a stat_*() function is probably easier to use when it comes to actually building a plot (and this might be reflected in the fact that geom_*() functions w stat defaults get the most direct usage.

**Comment by teunbrand** [comment]
I might have missed why you want to hardcode the stat/geom parts of a layer, could you explain this a bit more?
To my mind, you can just make the desired geom/stat the default in the formals of the constructor function, so you get flexibility and the correct behaviour you want. 

> It's actually happened to me with ggbump::geom_sigmoid.

I definitely also bumped into this, but with the stat :)

**Reply by EvaMaeRey** [reply]
> I definitely also bumped into this, but with the stat :)  

I wanted to do annotate(geom = "sigmoid" .....)!  But I think StatSigmoid might be just at work with GeomLine.  Now I'm curious about ggbump's naming conventions. 

I just describe some cases that might help at what I'm getting at in response to June's reply.

**Comment by yjunechoe** [comment]
>I think this kind of runs counter to a hadley quote that having stat_ and geom_ functions was the wrong way to go ...

The quote is from the "Extending ggplot" vignette!

> Next we write a layer function. Unfortunately, due to an early design mistake I called these either stat_() or geom_(). A better decision would have been to call them layer_() functions: that’s a more accurate description because every layer involves a stat and a geom.

https://ggplot2.tidyverse.org/articles/extending-ggplot2.html

**Reply by EvaMaeRey** [reply]
> 'early design mistake' 

Ah, the vignette!  Thanks, @yjunechoe!!

I wonder if it ggplot2 would feel so easy to use if you had specify geom and stat.  Maybe that isn't the counterfactual.  Maybe it's layer_point_identity() user-facers?  

I kind have the following pattern in some work:

geom_circlepack (for the 'main' anticipated use w/ GeomPolygon)
geom_circlepack_text  - and actually the Stat isn't even the same as the above, because labels are at the center... 😬

I think the likert usage in ggstats names similarly. 

Access to composability is awesome but maybe the conventions actually used might make things more fluid and concise for the user?

Based on a compute_panel_state that adds on geometry column of perimeters based on aes(state_name = my_state_name_var).  I'm following this pattern...

geom_states()   - GeomSf, StatState
geom_states_label() - GeomLabel, StatState
geom_states_text()  - GeomText, StatState

and stat specified...

stat_states()  - StatState with default GeomPoint

I think I've seen the stat_*/ geom_* aliasing with geom_smooth() but it seems much more careful...

@yjunechoe  commented on my lazy aliasing in the recipes:

>  (Opinionated) On Step 3, when defining geom_median(), it may be kinda odd to let the layer's geom be swappable in a geom_median(geom = ...). Not really a constructive feedback but something to keep in mind (I doubt anyone will actually try to pass a geom to a geom_*() function)

I think people like to use geom_* (given current naming conventions) because they can associate it with something they can see and stat_* is like something that is performed, and isn't as much a description of the plot you envision.  geom_point use >> stat_identity use

**Reply by yjunechoe** [reply]
> I wonder if it ggplot2 would feel so easy to use if you had specify geom and stat. Maybe that isn't the counterfactual. Maybe it's layer_point_identity() user-facers?

I think the "problem" of `stat_`/`geom_` vs. `layer_` is more apparent for cases like the existing `stat_boxplot()`/`geom_boxplot()` constructors with `StatBoxplot` and `GeomBoxplot` ggprotos. Under the current design, it's very difficult for a user to isolate the contribution of the geom and the stat and learn about their roles in a layer. So for at least that extreme (though not too rare - seealso Geom/StatSmooth) example, I'd think the counterfactual would be:

```
layer_boxplot <- function(...) {
  layer(stat = StatFiveNum, geom = GeomBoxesWhiskers, ...)
}
```

> I think people like to use geom_* (given current naming conventions) because they can associate it with something they can see and stat_* is like something that is performed, and isn't as much a description of the plot you envision. geom_point use >> stat_identity use

I think that speaks exactly to hadley's point about this perhaps being a "design mistake" - it makes users lose sight of the fact that stats and geoms are both layers, and every layer has both a stat and a geom. That in turn creates a larger barrier to entry into the internals, when you're forced to unlearn the intuition that `geom_*()`s "draw stuff on the canvas" and `stat_*()`s "compute stuff in the background". I realize that my position is less friendly, but I think it'll serve users better in the long run to unlearn this and get comfortable using `stat_()` layers and play around with mixing up different stats and geoms within in a layer.

**Reply by EvaMaeRey** [reply]
Yes, the layers that Geom/Stats with greater complementarity sound particularly illuminating.


Mixing up is good practice... but for the plot that needs to be specified rather quickly, maybe fully dedicated layer has a role.

I think I'm comfortable a pattern like this:

1. stat_county   # uses GeomSf as a default, uses Stat county computation
2. geom_county_sf # hard code GeomSf, otherwise same as stat_county - I didn't even bother with this name before this conversation so... moving the needle
3. geom_county_label # hard code GeomLabel, otherwise same as stat_county
4. geom_county_text # hard code to GeomText, otherwise same as stat_county

And finally, which I can see as most objectionable, short version of layer we anticipate being used most, geom_county_sf()

5. geom_county <- geom_county_sf    # aliasing

I'm attached to this last name;  presenting as a shortened version of geom_county_sf seems less problematic somehow!

All except the last seem pretty close to the layer_stat_geom ideal, in the end?  I think stat name comes first for user concerns.  

Is it friendly and principled?

**Reply by yjunechoe** [reply]
> Mixing up is good practice... but for the plot that needs to be specified rather quickly, maybe fully dedicated layer has a role.

Of course - everything in moderation!

> Is it friendly and principled?

I don't work with maps and sf at all, but the pairing of stats and geoms make good sense to me. I'm not sure if by hard-coding you mean (unable to be changed by the user) but I think I typically prefer things to be default values instead (sharing Teun's point; I'm not as good at articulating this position 😅).

**Reply by EvaMaeRey** [reply]
In sf2stat, a routine that creates many dedicated layers (with Stat and Geom fixed) using the geom_* prefix in it's 'layer' sense, is now demonstrated - currently with @emilyriederer's NC electoral data extract.  Very nice, and super-fitting since ggplot's SF examples are already with NC counties, https://ggplot2.tidyverse.org/reference/ggsf.html#ref-examples

https://github.com/EvaMaeRey/sf2stat?tab=readme-ov-file#sf2stat-proposed-usage

First create a kind of proper stat (though StatCounty won't exist and StatRegion also doesn't)

stat_county <- function(...){stat_region(ref_data = nc_ref, required_aes = "county_name|fips", id_index = 1, ...)}  

stat_region GeomSf as default geom and sf2stat::StatRegion as stat.  stat_region's compute_panel has a geographic ref_data argument and id_index (for specifying which id to do labeling with if there are multiple).  Internally, a temporary Stat is created StatSfJoin is created, and you can change the required_aes from the stat_region() function to remind users what geographic ids can be used so the regions are added on and plotted. 

Routine to create convenience layers...

geom_county_sf <- function(...){stat_county(geom = GeomSf,...)}
geom_county <- geom_county_sf   # convenience short name
geom_county_outline <- function(...){stat_county(geom = GeomOutline, ...)}  # GeomOutline is a default_aes-modified GeomSf
geom_county_label <- function(...){stat_county(geom = GeomLabel,...)}
geom_county_text <- function(...){stat_county(geom = GeomText, ...)}

And annotation convenience layers, w/ stamp_ prefix:  😬

stamp_county_sf <- function(...){geom_county_sf(stamp = T, ...)}
stamp_county <- stamp_county_sf
stamp_county_outline <- function(...){geom_county_outline(stamp = T, ...)}
stamp_county_label <- function(...){geom_county_label(stamp = T, ...)}
stamp_county_text <- function(...){geom_county_text(stamp = T, ...)}


Then you can do something like this...

read.csv("nc-midterms.csv") |>
  ggplot() + 
  aes(county_name = str_to_title(desc_county)) + 
  stamp_county(fill = 'darkgrey') +              # base layer with all counties displayed
  geom_county() + aes(fill = n/100000) +  # choropleth layer
  stamp_county_outline(keep_id = "Mecklenburg",  color = "orange",  linewidth = 1) + # outlining Mecklenburg
  geom_county_text(color = "white",   check_overlap = T, size = 2) # Addling annotation (defaults to column one of ref data, county_name)

![image](https://github.com/user-attachments/assets/f78b1461-83c2-4861-9ac1-0178df184b10)

**Reply by yjunechoe** [reply]
For my Q to Hadley today. I tracked down the "early design mistake" quote to the first time it appeared on GitHub. It's in a commit on August 2015 and there's a comment on that line extrapolating on it a bit too! https://github.com/tidyverse/ggplot2/commit/2a3e394902fa7ccac641d50a5ac4711152b24764#diff-2d9aa7a5e6a251ffeefa991e2153651ffc71e68086d9844bbe54ba65877ea5a0R68
